pyemma/coordinates/transform/transformer.py

```diff
@@ -35,15 +35,89 @@
 __all__ = ['Transformer']
 __author__ = 'noe, marscher'
 
+class TransformerIteratorContext(object):
+
+    def __init__(self, stride=1, lag=0):
+        self._lag = lag
+        self.__init_stride(stride)
+
+    def __init_stride(self, stride):
+        self._stride = stride
+        if isinstance(stride, np.ndarray):
+            keys = stride[:, 0]
+            self._trajectory_keys, self._trajectory_lengths = np.unique(keys, return_counts=True)
+        else:
+            self._trajectory_keys = None
+        self._uniform_stride = TransformerIteratorContext.is_uniform_stride(stride)
+        if not self.uniform_stride and not self.is_stride_sorted():
+            raise ValueError("Currently only sorted arrays allowed for random access")
+
+    def ra_indices_for_traj(self, traj):
+        """
+        Gives the indices for a trajectory file (without changing the order within the trajectory itself).
+        :param traj:
+        :return:
+        """
+        assert not self.uniform_stride, "requested random access indices, but is in uniform stride mode"
+        return self._stride[self._stride[:, 0] == traj][:, 1] if traj in self.traj_keys else np.array([])
```

Comment:
Alright, I have run some tests now. The speed **strongly** depends on the file format which is used. For the file formats that were faster i have randomly selected 10.000 samples from 50.000 frames and for the slower formats 1.000 from 10.000 frames.

1. 10.000 samples from 50.000 frames:

Format | Frames from files | Random access Stride
------ | --------------- | --------------------
h5 | 83 seconds | 21 seconds
dcd | 86 seconds | 3 seconds
binpos | 86 seconds | 2 seconds
nc | 86 seconds | 3 seconds

2. 1.000 samples from 10.000 frames:

Format | Frames from files | Random access Stride
------ | --------------- | --------------------
h5 | 1.76 seconds | 2.08 seconds
dcd | 1.62 seconds | 0.31 seconds
binpos | 1.57 seconds | 0.17 seconds
nc | 1.62 seconds | 0.33 seconds
xtc | 1.65 seconds | 23.58 seconds
trr | 1.73 seconds | 52.45 seconds

... as you can see, xtc and trr are quite bad formats for the 'seek-whence'-approach. I have found this issue rmcgibbo/mdtraj#336, which suggests that xtc's seek is not quite efficient.